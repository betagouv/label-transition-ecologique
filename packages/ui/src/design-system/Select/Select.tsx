import { SelectBase, SelectProps } from './components/SelectBase';

/**
 * SÃ©lecteur de valeur unique
 *
 * Ajouter `isSearcheable`, `createProps` ou `onSearch` pour faire un Searchable select
 */
export const Select = ({ values, onChange, ...props }: SelectProps) => {
  return (
    <SelectBase
      {...props}
      //values sera toujours un OptionValue simple sans tableau
      onChange={(v) => onChange(v !== values ? v : '')}
      values={values}
    />
  );
};
