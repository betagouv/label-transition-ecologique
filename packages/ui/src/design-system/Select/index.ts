export * from './Select';
export * from './SelectMultiple';
export * from './SelectFilter';
