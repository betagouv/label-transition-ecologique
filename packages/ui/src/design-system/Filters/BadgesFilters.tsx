import { Badge } from '@tet/ui/design-system/Badge';
import { ButtonMenu } from '@tet/ui/design-system/Button';
import { Field } from '@tet/ui/design-system/Field';
import {
  OptionValue,
  Select,
  SelectMultiple,
  SelectMultipleOnChangeArgs,
  SelectOption,
  getFlatOptions,
} from '@tet/ui/design-system/Select';
import classNames from 'classnames';
import { useEffect, useState } from 'react';

type FilterType = {
  /** Titre affiché dans <Field /> */
  title: string;
  /** Tag optionnel, affiché dans les badges */
  tag?: string;
  /** Liste des options du select */
  options: SelectOption[];
  /** Valeurs par défaut du select */
  values: OptionValue | OptionValue[] | undefined;
  /** Active la multi sélection dans les filtres */
  multiple?: boolean;
  /** Détecte le changement de valeur du select */
  onChange: (args?: SelectMultipleOnChangeArgs | OptionValue) => void;
};

type BadgeType = {
  filter: FilterType;
  value: OptionValue;
  label: string;
};

type FiltersMenuProps = {
  filters: FilterType[];
  className?: string;
  /** Pour styler le bouton */
  btnMenuClassName?: string;
  /** ClassName pour le container du NotificationButton */
  notificationBtnClassName?: string;
};

export const BadgesFilters = ({
  filters,
  className,
  btnMenuClassName,
  notificationBtnClassName,
}: FiltersMenuProps) => {
  const [badgesList, setBadgesList] = useState<BadgeType[] | null>(null);

  /** Gère la fermeture d'un badge et la mise à jour du filtre associé */
  const handleCloseBadge = (badge: BadgeType) => {
    if (Array.isArray(badge.filter.values)) {
      const selectedValue = badge.value;

      const newValuesArray = badge.filter.values?.filter(
        (v) => v !== badge.value
      );

      const values =
        newValuesArray && newValuesArray.length > 0
          ? newValuesArray
          : undefined;

      badge.filter.onChange({ selectedValue, values });
    } else {
      badge.filter.onChange();
    }
  };

  /** Supprime tous les filtres sélectionnés */
  const handleClearFilters = () => {
    badgesList?.forEach((badge) => {
      badge.filter.multiple
        ? badge.filter.onChange({
            selectedValue: badge.value,
            values: undefined,
          })
        : badge.filter.onChange();
    });
  };

  /** Actualise la liste de badge lors de la mise à jour des valeurs des filtres */
  useEffect(() => {
    const newList: BadgeType[] = [];

    filters.forEach((filter) => {
      if (filter.values !== undefined) {
        const options = getFlatOptions(filter.options);

        if (Array.isArray(filter.values)) {
          filter.values.forEach((value) => {
            const option = options.find((opt) => opt.value === value);
            if (!option) return;

            newList.push({
              filter,
              value: option.value,
              label: option.label,
            });
          });
        } else {
          const option = options.find((opt) => opt.value === filter.values);
          if (!option) return;

          newList.push({
            filter,
            value: option.value,
            label: option.label,
          });
        }
      }
    });

    newList.length > 0 ? setBadgesList(newList) : setBadgesList(null);
  }, [JSON.stringify(filters.map((f) => f.values))]);

  return (
    <div
      className={classNames('flex justify-between items-start', className)}
      title="Filtrer"
    >
      <div className="flex gap-2 flex-wrap">
        {/* Liste des badges correspondant aux filtres sélectionnés */}
        {!!badgesList &&
          badgesList.map((badge) => (
            <Badge
              key={`${badge.label}-${badge.value}`}
              title={`${badge.filter.tag ? `${badge.filter.tag} : ` : ''}${
                badge.label
              }`}
              size="sm"
              state="standard"
              uppercase={false}
              light
              onClose={() => handleCloseBadge(badge)}
            />
          ))}

        {/* Badge pour effacer tous les filtres sélectionnés */}
        {!!badgesList && badgesList.length > 0 && (
          <div onClick={handleClearFilters} className="cursor-pointer">
            <Badge
              title="Supprimer tous les filtres"
              icon="delete-bin-6-line"
              iconPosition="left"
              size="sm"
              state="grey"
              uppercase={false}
              light
              onClose={() => undefined}
            />
          </div>
        )}
      </div>

      {/* Menu + Filtres Select */}
      <ButtonMenu
        icon="equalizer-fill"
        notificationValue={badgesList?.length}
        className={btnMenuClassName}
        notificationBtnClassName={notificationBtnClassName}
      >
        <div className="flex flex-col gap-4 w-72">
          {filters.map((filter) => (
            <Field key={filter.title} title={filter.title}>
              {filter.multiple ? (
                <SelectMultiple
                  options={filter.options}
                  values={Array.isArray(filter.values) ? filter.values : []}
                  onChange={filter.onChange}
                  small
                />
              ) : (
                <Select
                  options={filter.options}
                  values={filter.values}
                  onChange={(args) => {
                    filter.onChange(args);
                  }}
                  small
                />
              )}
            </Field>
          ))}
        </div>
      </ButtonMenu>
    </div>
  );
};
