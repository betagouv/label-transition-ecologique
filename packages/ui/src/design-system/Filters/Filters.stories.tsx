import {Meta, StoryObj} from '@storybook/react';
import {BadgesFilters} from './BadgesFilters';
import {useState} from 'react';
import {OptionValue, SelectMultipleOnChangeArgs} from '@design-system/Select';

const meta: Meta<typeof BadgesFilters> = {
  component: BadgesFilters,
};

export default meta;

type Story = StoryObj<typeof BadgesFilters>;

export const FiltersWithBadges: Story = {
  render: () => {
    const [options1, setOptions1] = useState<OptionValue | undefined>();
    const [options2, setOptions2] = useState<OptionValue[] | undefined>();
    const [options3, setOptions3] = useState<OptionValue[] | undefined>();

    return (
      <div className="mb-[500px]">
        <BadgesFilters
          filters={[
            {
              title: 'Option 1',
              options: [
                {value: 'option11', label: 'Option 1.1'},
                {value: 'option12', label: 'Option 1.2'},
                {
                  value: 'option13',
                  label: 'Option 1.3',
                },
              ],
              values: options1,
              onChange: (args: OptionValue) => setOptions1(args),
            },
            {
              title: 'Option 2',
              tag: 'Opt 2',
              options: [
                {value: 'option21', label: 'Option 2.1'},
                {value: 'option22', label: 'Option 2.2'},
                {
                  value: 'option23',
                  label: 'Option 2.3',
                },
              ],
              values: options2,
              onChange: (args: SelectMultipleOnChangeArgs) =>
                setOptions2(args.values),
              multiple: true,
            },
            {
              title: 'Option 3',
              tag: 'Opt 3',
              options: [
                {
                  title: 'Option 3.1',
                  options: [
                    {value: 'option311', label: 'Option 3.1.1'},
                    {value: 'option312', label: 'Option 3.1.2'},
                  ],
                },
                {
                  title: 'Option 3.2',
                  options: [
                    {value: 'option321', label: 'Option 3.2.1'},
                    {value: 'option322', label: 'Option 3.2.2'},
                  ],
                },
                {
                  title: 'Option 3.3',
                  options: [
                    {value: 'option331', label: 'Option 3.3.1'},
                    {value: 'option332', label: 'Option 3.3.2'},
                  ],
                },
              ],
              values: options3,
              onChange: (args: SelectMultipleOnChangeArgs) =>
                setOptions3(args.values),
              multiple: true,
            },
          ]}
        />
      </div>
    );
  },
};
