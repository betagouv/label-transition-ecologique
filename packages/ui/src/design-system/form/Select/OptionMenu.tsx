import {Ref, forwardRef} from 'react';
import DropdownFloater from '../../../components/floating-ui/DropdownFloater';

import {CreateOption} from './Select';
import {Option} from './Options';
import {Icon} from '../../icons/Icon';
import {Button} from '../../buttons/button/Button';

type Props = {
  option: Option;
  createProps: CreateOption;
};

export const OptionMenu = ({option, createProps}: Props) => (
  <DropdownFloater
    offsetValue={{mainAxis: 6}}
    placement="top-end"
    noDropdownStyles
    render={() => (
      <div
        className="flex flex-col divide-y divide-x-0 divide-solid divide-grey-3 rounded overflow-hidden bg-white border border-grey-3"
        onClick={evt => {
          evt.stopPropagation();
        }}
      >
        <button
          className="flex items-center w-full py-2 pr-4 pl-3 text-sm text-grey-8"
          onClick={() => createProps.onUpdate(option.value, 'TODO')}
        >
          <Icon icon="edit-line" size="sm" className="mr-2" />
          Ã‰diter
        </button>
        <button
          className="flex items-center w-full py-2 pr-4 pl-3 text-sm text-grey-8"
          onClick={() => createProps.onDelete(option.value)}
        >
          <Icon icon="delete-bin-6-line" size="sm" className="mr-2" />
          Supprimer
        </button>
      </div>
    )}
  >
    <OptionMenuButton />
  </DropdownFloater>
);

/** Bouton pour ouvrir le menu d'une option */
const OptionMenuButton = forwardRef((props, ref?: Ref<HTMLButtonElement>) => (
  <div className="flex px-3 py-1.5 my-auto group-hover:bg-primary-0 hover:!bg-white">
    <Button
      ref={ref}
      icon="more-line"
      size="xs"
      variant="white"
      className="!p-1 m-auto"
      {...props}
    />
  </div>
));
OptionMenuButton.displayName = 'OptionOpenFloaterButton';
