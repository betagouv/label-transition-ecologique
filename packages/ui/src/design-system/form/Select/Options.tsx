import {Fragment} from 'react';
import classNames from 'classnames';

import {isOptionSection} from './utils';
import {OptionMenu} from './OptionMenu';
import {CreateOption} from './Select';
import {Badge} from '../../badge/Badge';

/**
 * Types partagés entre tous les composants selects
 * (Select, MultiSelect, MultiSelectFilter)
 */

/** Type du champ valeur d'une option */
export type OptionValue = number | string;
/** Type de base d'une option générique */
export type Option = {value: OptionValue; label: string};
/** Type d'une liste d'options dans un sélecteur */
export type OptionSection = {
  title: string;
  options: Option[];
};
/** Type d'une option dans un sélecteur, peut être une simple option ou une liste d'options */
export type SelectOption = Option | OptionSection;

type BaseProps<T extends OptionValue> = {
  /** Liste des valeurs sélectionnées dans le sélecteur parent */
  values?: T[];
  /** Appelée au click d'une option (reçoit la valeur de l'option cliquée) */
  onChange: (value: T) => void;
  /** Permet de customiser l'item (label) d'une option */
  customItem?: (option: Option) => React.ReactElement;
  /** Les fonction permettant la création de nouvelles options */
  createProps?: CreateOption;
};

type OptionsListProps<T extends OptionValue> = BaseProps<T> & {
  /** Id pour les tests e2e */
  dataTest?: string;
  /** Liste des options */
  options: SelectOption[];
  /** Fait apparaître un état de chargement à la place des options */
  isLoading: boolean;
  /** Texte affiché quand il n'y a aucune option fournie */
  noOptionPlaceholder?: string;
};

/** Liste d'options pouvant être de simples options ou des sections */
const Options = <T extends OptionValue>({
  values,
  options,
  onChange,
  isLoading,
  customItem,
  createProps,
  noOptionPlaceholder,
  dataTest,
}: OptionsListProps<T>) => {
  return (
    <div data-test={`${dataTest}-options`}>
      {isLoading ? (
        <div className="p-4 text-sm text-gray-500">Chargement...</div>
      ) : options.length > 0 ? (
        options.map((option, i) => {
          /** Section */
          if (isOptionSection(option)) {
            return (
              <Fragment key={option.title + i}>
                <div className="first:hidden h-[1px] mx-6 bg-grey-4" />
                <div>
                  {/** titre */}
                  <div className="pt-4 pb-2 mx-6 font-bold text-left text-sm uppercase text-primary-7">
                    {option.title}
                  </div>
                  {/** options */}
                  {option.options.map(option => (
                    <Option
                      key={option.value}
                      option={option}
                      values={values}
                      onChange={onChange}
                      customItem={customItem}
                      createProps={createProps}
                    />
                  ))}
                </div>
              </Fragment>
            );
            /** Option simple */
          } else {
            return (
              <Option
                key={option.value}
                option={option}
                values={values}
                onChange={onChange}
                customItem={customItem}
                createProps={createProps}
              />
            );
          }
        })
      ) : (
        <div className="py-4 px-6 text-sm text-gray-500">
          {noOptionPlaceholder || 'Aucune option disponible'}
        </div>
      )}
    </div>
  );
};

export default Options;

type OptionProps<T extends OptionValue> = BaseProps<T> & {
  option: Option;
};

/** Option pour les sélecteurs */
const Option = <T extends OptionValue>({
  values,
  option,
  onChange,
  customItem,
  createProps,
}: OptionProps<T>) => {
  const isActive = values?.includes(option.value as T);
  const isUserCreated = createProps?.userCreatedOptions.includes(option.value);
  return (
    <button
      data-test={option.value}
      className="flex items-start w-full p-2 pr-6 text-left text-sm hover:!bg-primary-0"
      onClick={() => onChange(option.value as T)}
    >
      <div className="flex w-6 mr-2 shrink-0">
        {isActive && (
          <span className="fr-icon-check-line flex mt-1 mx-auto text-primary-7 before:m-auto before:!h-4 before:!w-4" />
        )}
      </div>
      <div className="flex mr-auto my-auto">
        {customItem ? (
          customItem(option)
        ) : createProps ? (
          <Badge
            title={option.label}
            state={isUserCreated ? 'standard' : 'default'}
            size="sm"
          />
        ) : (
          <span
            className={classNames('leading-6 text-grey-8', {
              'text-primary-7': isActive,
            })}
          >
            {option.label}
          </span>
        )}
      </div>
      {createProps &&
        createProps.userCreatedOptions.some(o => o === option.value) && (
          <OptionMenu option={option} createProps={createProps} />
        )}
    </button>
  );
};
