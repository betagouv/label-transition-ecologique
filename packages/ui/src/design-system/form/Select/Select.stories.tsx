import {useState} from 'react';
import {Meta, StoryObj} from '@storybook/react';
import {action} from '@storybook/addon-actions';

import {Field} from '@design-system/Field';

import {Select} from './Select';
import {OptionValue, SelectOption} from './Options';
import {getFlatOptions, onSelectMultiple, onSelectSingle} from './utils';

const singleOptions: SelectOption[] = [
  {value: 'option1', label: 'Option 1'},
  {value: 'option2', label: 'Option 2'},
  {
    value: 'option3',
    label:
      "Un titre d'option super long, car nous avons besoin de voir le comportement",
  },
];

const optionsWithSections: SelectOption[] = [
  {value: 'ss1', label: 'Option sans section 1'},
  {value: 'ss2', label: 'Option sans section 2'},
  {
    title: 'Section 1',
    options: [
      {value: 'option1', label: 'Option 1'},
      {value: 'option2', label: 'Option 2'},
      {
        value: 'option3',
        label:
          "Un titre d'option super long, car nous avons besoin de voir le comportement",
      },
    ],
  },
  {
    title: 'Section 2',
    options: [
      {value: 'option4', label: 'Option 4'},
      {value: 'option5', label: 'Option 5'},
    ],
  },
];

const meta: Meta<typeof Select> = {
  component: Select,
  decorators: [story => <div className="w-full max-w-[24rem]">{story()}</div>],
};

export default meta;

type Story = StoryObj<typeof Select>;

export const Default: Story = {
  args: {options: singleOptions},
  render: args => {
    const [value, setValue] = useState<OptionValue | undefined>();
    return (
      <Select
        {...args}
        values={value}
        onChange={v => {
          setValue(onSelectSingle(v, value));
          action('onChange');
        }}
      />
    );
  },
};

export const SmallWithBadgeItems: Story = {
  args: {options: singleOptions, small: true, isBadgeItem: true},
  render: args => {
    const [value, setValue] = useState<OptionValue | undefined>();
    return (
      <Select
        {...args}
        values={value}
        onChange={v => {
          setValue(onSelectSingle(v, value));
          action('onChange');
        }}
      />
    );
  },
};

export const Disabled: Story = {
  args: {options: singleOptions, onChange: () => null, disabled: true},
};

export const CustomItem: Story = {
  args: {options: singleOptions},
  render: args => {
    const [value, setValue] = useState<OptionValue | undefined>();
    return (
      <Select
        {...args}
        values={value}
        onChange={v => {
          setValue(onSelectSingle(v, value));
          action('onChange');
        }}
        customItem={option => (
          <div className="flex items-center px-2 py-1 text-sm text-white bg-warning-1 rounded-xl">
            {option.label}
          </div>
        )}
      />
    );
  },
};

export const MultiSelectWithSectionOptions: Story = {
  args: {options: optionsWithSections, multiple: true},
  render: args => {
    const [values, setValues] = useState<OptionValue[] | undefined>();
    return (
      <Select
        {...args}
        values={values}
        onChange={v => {
          setValues(onSelectMultiple(v, values));
        }}
      />
    );
  },
};

export const SearchableMultiSelect: Story = {
  args: {
    options: optionsWithSections,
    multiple: true,
    isSearcheable: true,
    emptySearchPlaceholder:
      "Placeholder custom pour la liste d'option sans résultat",
  },
  render: args => {
    const [values, setValues] = useState<OptionValue[] | undefined>();
    return (
      <Select
        {...args}
        values={values}
        onChange={v => {
          setValues(onSelectMultiple(v, values));
        }}
      />
    );
  },
};

export const DisabledSearchableMultiSelectWithValue: Story = {
  args: {
    options: singleOptions,
    disabled: true,
    isSearcheable: true,
    multiple: true,
  },
  render: args => {
    const [values, setValues] = useState<OptionValue[] | undefined>([
      'option1',
      'option2',
    ]);
    return (
      <Select
        {...args}
        values={values}
        onChange={v => {
          setValues(onSelectMultiple(v, values));
        }}
      />
    );
  },
};

export const SearchableSelectWithDebouncedApiCallOnTyping: Story = {
  args: {
    debounce: 500,
    options: singleOptions,
  },
  render: args => {
    const [value, setValue] = useState<OptionValue | undefined>();
    return (
      <Select
        {...args}
        values={value}
        onChange={v => {
          setValue(onSelectSingle(v, value));
        }}
        onSearch={v => console.log(v)}
      />
    );
  },
};

export const CreateOption: Story = {
  args: {
    multiple: true,
  },
  render: args => {
    const [values, setValues] = useState<OptionValue[] | undefined>();
    const [options, setOptions] = useState(singleOptions);
    const userCreatedOptions = getFlatOptions(options)
      .filter(o => typeof o.value === 'number')
      .map(o => o.value);

    return (
      <Select
        {...args}
        options={options}
        values={values}
        onChange={v => {
          setValues(onSelectMultiple(v, values));
        }}
        createProps={{
          userCreatedOptions,
          onCreate: label => {
            const newOption: SelectOption = {
              label,
              value: Date.now(),
            };
            setOptions([...options, newOption]);
            setValues(onSelectMultiple(newOption.value, values));
          },
          onUpdate: (value, label) =>
            setOptions(
              getFlatOptions(options).map(o =>
                o.value !== value ? o : {label, value}
              )
            ),
          onDelete: value => {
            setOptions(getFlatOptions(options).filter(o => o.value !== value));
            setValues(onSelectMultiple(value, values));
          },
        }}
      />
    );
  },
};

export const WithField: Story = {
  args: {options: singleOptions},
  render: args => {
    const [value, setValue] = useState<OptionValue | undefined>();
    return (
      <Field
        title="Description de l'action"
        hint="Texte description additionnel"
        state="info"
        message="Message d’information"
      >
        <Select
          {...args}
          values={value}
          onChange={v => {
            setValue(onSelectSingle(v, value));
            action('onChange');
          }}
        />
      </Field>
    );
  },
};
