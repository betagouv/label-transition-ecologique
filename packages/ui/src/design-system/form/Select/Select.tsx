import {Placement} from '@floating-ui/react';
import {Ref, forwardRef, useEffect, useState} from 'react';
import classNames from 'classnames';
import {useDebouncedCallback} from 'use-debounce';

import DropdownFloater from '../../../components/floating-ui/DropdownFloater';
import Options, {OptionValue, SelectOption, Option} from './Options';
import {Badge} from '../../badge/Badge';

import {
  filterOptions,
  getOptionLabel,
  getFlatOptions,
  sortOptionByAlphabet,
} from './utils';
import {Icon} from '../../icons/Icon';

export type CreateOption = {
  userCreatedOptions: OptionValue[];
  onCreate: (inputValue: string) => void;
  onDelete?: (id: OptionValue) => void;
  onUpdate?: (id: OptionValue, inputValue: string) => void;
};

type SelectProps<T extends OptionValue> = {
  /** Id pour les tests e2e */
  dataTest?: string;
  /** Liste des options */
  options: Array<SelectOption>;
  /** Appelée au click d'une option (reçoit la valeur de l'option cliquée) */
  onChange: (value: T) => void;
  /** Valeurs sélectionnées, peut recevoir une valeur seule ou un tableau de valeurs */
  values?: T | T[];
  /** Active la multi sélection */
  multiple?: boolean;
  /** Permet la recherche dans la liste d'option */
  isSearcheable?: boolean;
  /** Fonction exécutée lorsque l'utilisateur fait une recherche, reçoit la valeur de l'input */
  onSearch?: (v: string) => void;
  /** Temps du debounce appliqué à onSearch */
  debounce?: number;
  /** Les propriétés permettant la création de nouvelles options */
  createProps?: CreateOption;
  /** Fait apparaître un état de chargement à la place des options */
  isLoading?: boolean;
  /** Permet de désactiver le bouton d'ouverture */
  disabled?: boolean;
  /** Texte affiché quand rien n'est sélectionné */
  placeholder?: string;
  /** Permet de customiser l'item (label) d'une option */
  customItem?: (option: Option) => React.ReactElement;
  isBadgeItem?: boolean;
  /** Texte affiché quand aucune option ne correspond à la recherche */
  emptySearchPlaceholder?: string;
  /** Change le positionnement du dropdown menu */
  placement?: Placement;
  /** Pour que la largeur des options ne dépasse pas la largeur du bouton d'ouverture */
  containerWidthMatchButton?: boolean;
  /** Affiche une version plus petite du sélecteur */
  small?: boolean;
};

/**
 * Composant multi-fonction:
 *
 * Select /
 * Multi select /
 * Searchable select /
 * Create option select
 *
 * Mettre `multiple` à `true` pour faire un multi select
 *
 * Ajouter `isSearcheable`, `createProps` ou `onSearch` pour faire un Searchable select
 *
 * Donner `createProps` pour faire un Create option select, cela active `isSearcheable' automatiquement
 */
export const Select = <T extends OptionValue>(props: SelectProps<T>) => {
  const {
    dataTest,
    values,
    options,
    onChange,
    createProps,
    onSearch,
    debounce = onSearch ? 250 : 0,
    placeholder,
    emptySearchPlaceholder,
    placement,
    multiple = false,
    isSearcheable = false,
    isLoading = false,
    customItem,
    isBadgeItem = false,
    containerWidthMatchButton = true,
    disabled = false,
    small = false,
  } = props;

  const hasSearch = isSearcheable || !!createProps || !!onSearch;

  /** Recherche textuelle locale car `onSearch` n'est pas obligatoire */
  const [inputValue, setInputValue] = useState('');

  /**
   * Permet de profiter du debounce de l'input et d'afficher un text de chargement
   * quand l'utilisateur est entrain de faire une saisie.
   */
  const [loading, setLoading] = useState(isLoading);
  // synchronise l'état de loading interne avec l'externe
  useEffect(() => setLoading(isLoading), [isLoading]);

  /** Fonction de debounce */
  const handleDebouncedInputChange = useDebouncedCallback(v => {
    onSearch(v);
    setLoading(false);
  }, debounce);

  /** Permet synchroniser les différents inputChange */
  const handleInputChange = (value: string) => {
    setInputValue(value);
    // uniquement si la fonction `onSearch` est donnée
    // on applique la fonction de `debounce`, par défaut le debounce est à 0
    if (onSearch) {
      // on active le loading sachant qu'on le désactive à la fin de la fct de debounce
      setLoading(true);
      handleDebouncedInputChange(value);
    }
  };

  /** Liste d'option filtrée par la saisie de l'utilisateur */
  const filteredOptions = filterOptions(options, inputValue);

  /** Détermine si le Select permet la création d'option */
  const isCreateOptionSelect = !!createProps;

  /** Compare la valeur de l'input de recherche avec la première optin de la liste
   * pour afficher le bouton de création d'une option */
  const isNotSimilar =
    inputValue.toLowerCase().trim() !==
    getFlatOptions(filteredOptions)[0]?.label.toLowerCase().trim();

  /** Transforme une valeur simple en tableau, qui est plus facile à traiter dans les sous composants */
  const arrayValues = values
    ? Array.isArray(values)
      ? values
      : [values]
    : undefined;

  return (
    <DropdownFloater
      placement={placement}
      offsetValue={0}
      containerWidthMatchButton={containerWidthMatchButton}
      noDropdownStyles
      enterToToggle={!hasSearch}
      disabled={disabled}
      render={({close}) => (
        <div
          data-test={`${dataTest}-options`}
          className="bg-white rounded-b-lg border border-grey-4 border-t-0 overflow-hidden"
        >
          {/** Bouton de création d'une option */}
          {isCreateOptionSelect &&
            inputValue.trim().length > 0 &&
            isNotSimilar && (
              <button
                data-test={`${dataTest}-create-option`}
                className="flex w-full p-2 pr-6 text-left text-sm hover:!bg-primary-0 overflow-hidden"
                onClick={() => {
                  createProps.onCreate(inputValue);
                  handleInputChange('');
                }}
              >
                <div className="flex w-6 shrink-0 mt-1 mr-2">
                  <Icon
                    icon="add-line"
                    size="sm"
                    className="m-auto text-primary-7"
                  />
                </div>
                <Badge
                  title={inputValue}
                  state="standard"
                  size="sm"
                  className="my-auto mr-auto"
                />
                <span className="mt-1 ml-6 font-medium text-grey-8">Créer</span>
              </button>
            )}
          {/** Liste des options */}
          <Options
            dataTest={dataTest}
            values={arrayValues}
            options={
              isCreateOptionSelect
                ? sortOptionByAlphabet(filteredOptions)
                : filteredOptions
            }
            onChange={value => {
              onChange(value);
              setInputValue('');
              if (!multiple) {
                close();
              }
            }}
            isLoading={loading}
            createProps={createProps}
            customItem={customItem}
            isBadgeItem={isBadgeItem}
            noOptionPlaceholder={emptySearchPlaceholder}
          />
        </div>
      )}
    >
      <SelectButton
        dataTest={dataTest}
        values={arrayValues}
        options={options}
        onChange={onChange}
        isSearcheable={hasSearch}
        inputValue={inputValue}
        onSearch={handleInputChange}
        createProps={createProps}
        multiple={multiple}
        customItem={customItem}
        placeholder={placeholder}
        disabled={disabled}
        small={small}
      />
    </DropdownFloater>
  );
};

type SelectButtonProps<T extends OptionValue> = SelectProps<T> & {
  /** Donné par le DropdownFloater */
  isOpen?: boolean;
  /** Valeur de la saisie */
  inputValue: string;
};

/* Création d'un composant séparé pour passer la ref du boutton au floater */
const SelectButton = forwardRef(
  <T extends OptionValue>(
    {
      dataTest,
      isOpen,
      options,
      values,
      onChange,
      inputValue,
      isSearcheable,
      onSearch,
      createProps,
      multiple,
      customItem,
      placeholder,
      disabled,
      small,
      ...props
    }: SelectButtonProps<T>,
    ref?: Ref<HTMLButtonElement>
  ) => {
    useEffect(() => {
      if (!isOpen) {
        onSearch('');
      }
    }, [isOpen]);

    return (
      <button
        ref={ref}
        data-test={dataTest}
        aria-expanded={isOpen}
        aria-label="ouvrir le menu"
        className={classNames(
          'w-full rounded-xl border border-solid border-grey-4 disabled:border-grey-3 bg-grey-1 hover:!bg-primary-0 disabled:hover:!bg-grey-1 overflow-hidden',
          {'rounded-b-none': isOpen}
        )}
        disabled={disabled}
        {...props}
      >
        <div
          className={classNames('flex px-4', {
            'min-h-[2.5rem] py-1': small,
            'min-h-[3rem] py-2': !small,
          })}
        >
          <div className="flex grow flex-wrap gap-2 mr-4">
            {values && Array.isArray(values) ? (
              /** Listes des valeurs sélectionnées */
              <div className="flex items-center gap-2 grow">
                {customItem ? (
                  // Item custom
                  customItem({
                    value: values[0],
                    label: getOptionLabel(values[0], getFlatOptions(options)),
                  })
                ) : (
                  // Badge par défaut
                  <Badge
                    state={
                      createProps &&
                      createProps.userCreatedOptions.includes(values[0])
                        ? 'standard'
                        : 'default'
                    }
                    title={getOptionLabel(values[0], getFlatOptions(options))}
                    onClose={!disabled && (() => onChange(values[0]))}
                  />
                )}
                {/** Nombre de valeurs sélectionnées supplémentaires */}
                {values.length > 1 && (
                  <Badge title={`+${values.length - 1}`} state="info" />
                )}
              </div>
            ) : (
              /** Si pas de valeur et que la recherche n'est pas activée, on affiche un placeholder */
              !isSearcheable && (
                <span
                  className={classNames(
                    'my-auto text-xs text-grey-6 line-clamp-1',
                    {
                      '!text-grey-5': disabled,
                    }
                  )}
                >
                  {placeholder ?? multiple
                    ? 'Sélectionner une ou plusieurs options'
                    : 'Sélectionner une option'}
                </span>
              )
            )}
            {
              // on affiche l'input si le sélecteur est désactivé et ne possède pas de valeur
              // afin d'afficher le placeholder de l'input
              isSearcheable && !(disabled && values) && (
                <input
                  type="text"
                  className={classNames(
                    'w-full text-sm outline-0 placeholder:text-grey-6',
                    {'py-1': values}
                  )}
                  value={inputValue}
                  onChange={e => onSearch(e.target.value)}
                  onClick={evt => isOpen && evt.stopPropagation()}
                  placeholder={placeholder ?? 'Rechercher par mots-clés'}
                  disabled={disabled}
                />
              )
            }
          </div>
          {/** Icône flèche d'ouverture */}
          <Icon
            icon="arrow-down-s-line"
            size="sm"
            className={classNames(
              'mt-2 ml-auto text-primary-9',
              {'rotate-180': isOpen},
              {'!text-grey-5': disabled}
            )}
          />
        </div>
      </button>
    );
  }
);
SelectButton.displayName = 'SelectButton';
