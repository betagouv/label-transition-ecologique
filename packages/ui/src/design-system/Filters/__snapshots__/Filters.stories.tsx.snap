// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`design-system/Filters FiltersWithBadges smoke-test 1`] = `
<div class="mb-[500px]">
  <div class="flex justify-between items-start"
       title="Filtrer"
  >
    <div class="flex gap-2 flex-wrap">
    </div>
    <button class="w-fit flex items-center border-solid group gap-1 cursor-pointer rounded-lg border font-bold text-white hover:!text-white bg-primary-7 hover:!bg-primary-8 border-primary-7 hover:!border-primary-8 p-2">
      <span class="ri-equalizer-fill font-normal text-center h-4 w-4 text-[1rem] leading-[1rem] fill-white group-hover:fill-white">
      </span>
    </button>
  </div>
</div>
`;
